# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/cookie/CookieService.cpp
# Commit: 20708fe16d2a
# Full Hash: 20708fe16d2a1f7e85b609bbfe286ec842b620b4
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2024-12-18 04:27:40
# Regressor Bug: 1915355
# File Overlap Count: 1
# Description:
#   Bug 1915355 - Part 5: Exempt third-party cookie blocking if the channel is on the exception list. r=cookie-reviewers,bvandersloot,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223384
# ==============================================================================

diff -r 7c4b003d1e48 -r 20708fe16d2a netwerk/cookie/CookieService.cpp
--- a/netwerk/cookie/CookieService.cpp	Tue Dec 17 21:44:37 2024 +0000
+++ b/netwerk/cookie/CookieService.cpp	Tue Dec 17 21:44:37 2024 +0000
@@ -586,7 +586,8 @@
     moreCookieToRead = cookieParser.Parse(
         baseDomain, requireHostMatch, cookieStatus, cookieHeader, dateHeader,
         true, isForeignAndNotAddon, mustBePartitioned,
-        storagePrincipalOriginAttributes.IsPrivateBrowsing());
+        storagePrincipalOriginAttributes.IsPrivateBrowsing(),
+        loadInfo->GetIsOn3PCBExceptionList());
 
     if (!cookieParser.ContainsCookie()) {
       continue;
@@ -853,6 +854,9 @@
 
   nsCOMPtr<nsIConsoleReportCollector> crc = do_QueryInterface(aChannel);
 
+  nsCOMPtr<nsILoadInfo> loadInfo = aChannel ? aChannel->LoadInfo() : nullptr;
+  const bool on3pcdException = loadInfo && loadInfo->GetIsOn3PCBExceptionList();
+
   for (const auto& attrs : aOriginAttrsList) {
     CookieStorage* storage = PickStorage(attrs);
 
@@ -980,13 +984,16 @@
       // Check if we need to block the cookie because of opt-in partitioning.
       // We will only allow partitioned cookies with "partitioned" attribution
       // if opt-in partitioning is enabled.
+      //
+      // Note: If the cookie is on the 3pcd exception list, we will include
+      // the cookie.
       if (aIsForeign && cookieJarSettings->GetPartitionForeign() &&
           (StaticPrefs::network_cookie_cookieBehavior_optInPartitioning() ||
            (attrs.IsPrivateBrowsing() &&
             StaticPrefs::
                 network_cookie_cookieBehavior_optInPartitioning_pbmode())) &&
           !(cookie->IsPartitioned() && cookie->RawIsPartitioned()) &&
-          !aStorageAccessPermissionGranted) {
+          !aStorageAccessPermissionGranted && !on3pcdException) {
         continue;
       }
 